### ISSUE TITLE
Shebang missing interpreter

### ISSUE THREAD

**Issue by Demayl** [virtualmin/virtualmin-gpl#1]
Current shebang **_#!/usr/local/bin/perl_** is not so portable ( in my case on CentOS 6.8 is missing ).
My proposal is to use **_#!/usr/bin/env perl_** or at least the good old **_#!/usr/bin/perl_**
So it can use first perl interpreter in path.

**Comment by jcameron** [comment]
You can just run `ln -s /usr/bin/perl /usr/local/bin/`

**Comment by Demayl** [comment]
Yes, i can, but the idea is to skip this step and make it portable. It's only a proposal.

**Comment by swelljoe** [comment]
I think `/usr/bin/env perl` is a good idea. It doesn't work on Windows, but we currently don't work on Windows without a lot of tweaking, anyway. env does work on every Linux/BSD I'm aware of.

Literally nobody has perl in /usr/local/bin, anymore. ;-) (OK, BSD does, but they recommend `/usr/bin/env perl` as the solution to that problem.)

As far as I can tell, this is considered current best practice for Perl, Python, etc. applications.

That said, I'm pretty sure it'll have to happen *everywhere*; across Webmin, Virtualmin, etc.. Having different perl versions called in different places can cause things to not work or work unpredictably. So, maybe it should wait until we're already breaking things in Webmin 2.0.

**Comment by jcameron** [comment]
Actually, it really doesn't matter what perl path is in those files (apart from miniserv.pl), as Webmin executes CGI scripts inside the same process rather than forking another copy of Perl.

We can switch to `/usr/bin/env perl` pretty easily though, assuming that `env` is available on all Unix variants.

**Comment by qooob** [comment]
Should I also fix those in the theme `*.cgi` files right now?

**Comment by swelljoe** [comment]
env has been around for many years.

But, now that I think about it, we may be introducing subtle problems with this. If the user has installed a non-standard Perl, and it's the one env picks up, and we rely on a handful of dependencies in our RPM/deb packages...those deps wouldn't end up being installed for the alternative Perl, and so, SSL may not work (unless they've installed Net::SSLeay manually), modules that need other non-core modules won't work, etc. We don't provide a cpan installable build (yet), so they won't have a way to auto-resolve those dependencies. 

Likewise, for the mail processing stack, there are some Perl scripts...some of them run as the user with the user's environment, and so might pick up a Perl that is missing something we need.

This might be something that needs to wait until we have a way for users of alternative Perls to install Webmin and all of its dependencies from CPAN. I've got vague aspirations in that direction, and have started putting all of my new Perl development for Virtualmin into CPAN modules, but, it'll be a long while before Webmin can go that route.

**Comment by swelljoe** [comment]
That said, we already have that problem by defaulting to /usr/local/bin. This guarantees that if there's a Perl in /usr/local, it won't have the dependencies Webmin installs via RPM or dpkg deps. I think /usr/local is worse than the /usr/bin or /usr/bin/env option.

**Comment by swelljoe** [comment]
Oh, nevermin..Webmin fixes that when installed via rpm or deb.

**Comment by jcameron** [comment]
Yeah, the RPM and Deb packages use /usr/bin/perl in all the .cgi and .pl scripts.

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
